pennylane/io/qualtran_io.py

```diff
@@ -0,0 +1,338 @@
+                                dtype=object,
+                            )
+                            soq_to_wires_len += total_elements
+                            in_quregs[soq.reg.name] = ascending_vals.reshape(
+                                (*soq.reg.shape, soq.reg.bitsize)
+                            )
+                        soq_to_wires[soq] = in_quregs[soq.reg.name][soq.idx]
+                    continue
+
+                in_quregs = {
+                    reg.name: np.empty((*reg.shape, reg.bitsize), dtype=object)
+                    for reg in binst.bloq.signature.lefts()
+                }
+                # The out_quregs inform us of the total # of wires in the circuit to account for
+                # wires that are split or allocated in the cbloq
+                out_quregs = {
+                    reg.name: np.empty((*reg.shape, reg.bitsize), dtype=object)
+                    for reg in binst.bloq.signature.rights()
+                }
+
+                for pred in pred_cxns:
+                    soq = pred.right
+                    soq_to_wires[soq] = soq_to_wires[pred.left]
+                    if isinstance(soq_to_wires[soq], list) and len(soq_to_wires[soq]) == 1:
+                        soq_to_wires[soq] = soq_to_wires[soq][0]
+                    in_quregs[soq.reg.name][soq.idx] = np.squeeze(soq_to_wires[soq])
+
+                for succ in succ_cxns:
+                    soq = succ.left
+                    if soq.reg.side == qt.Side.RIGHT:
+                        # When in_quregs != out_quregs, it means that there are wires unaccounted
+                        # for. We account for these wires and update soq_to_wires and in_quregs
+                        # accordingly.
+                        if len(in_quregs) != len(out_quregs):
+                            total_elements = np.prod(soq.reg.shape) * soq.reg.bitsize
+                            ascending_vals = np.arange(
+                                soq_to_wires_len,
+                                total_elements + soq_to_wires_len,
+                                dtype=object,
+                            )
+                            soq_to_wires_len += total_elements
+                            in_quregs[soq.reg.name] = ascending_vals.reshape(
+                                (*soq.reg.shape, soq.reg.bitsize)
+                            )
+                        soq_to_wires[soq] = in_quregs[soq.reg.name][soq.idx]
+
+                total_wires = [int(w) for ws in in_quregs.values() for w in list(ws.ravel())]
+                mapped_wires = [wires[idx] for idx in total_wires if idx < len(wires)]
+                ghost_wires = [f"alloc_free_{val}" for val in total_wires if val >= len(wires)]
+                op = binst.bloq.as_pl_op(mapped_wires + ghost_wires)
+
+                if op:
+                    ops.append(op)
+        except (qt.DecomposeNotImplementedError, qt.DecomposeTypeError):
+            pass
```

Comment:
Could we instead raise a `DecompositionUndefinedError` from the `DecomposeNotImplementedError`?

Reply:
Do you mean to raise an error inside the except statement?